File: 205.isomorphic-strings.py
#
# @lc app=leetcode id=205 lang=python3
#
# [205] Isomorphic Strings
#

# @lc code=start

#By str process
# class Solution:
#     def isIsomorphic(self, s: str, t: str) -> bool:
#         return all(s.index(s[i]) == t.index(t[i])  for i in range(len(s)))

#這題目標只有s不要多對一
#By hashtable
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        lens = len(s)
        lent = len(t)
        if lens!=lent:
            return False
        hashtable1 = dict()
        #如果要用hashtable回頭找時建的
        # hashtable2 = dict()
        for i in range(lens):
            if s[i] in hashtable1:
                if hashtable1[s[i]]!=t[i]:
                    return False
            else:
                hashtable1[s[i]] = t[i]
                #每次有新的s[i]要加入時
                #回頭看有沒有i以前的s對應到同樣的t
                #可以用for迴圈回頭找或建另一個hashtable找
                #每次用for找,比較快?
                for j in range(i):
                    if hashtable1[s[j]]==t[i]:
                        return False            
                #用hash table找
                # if t[i] in hashtable2:
                #     if hashtable2[t[i]]!=s[i]:
                #         return False
                # hashtable2[t[i]] = s[i]
        return True

# @lc code=end

File: 205.isomorphic-strings.c
<<<<<<< HEAD
/*
 * @lc app=leetcode id=205 lang=c
 *
 * [205] Isomorphic Strings
 */

// @lc code=start

// By hashtable
struct HashTable {
    char key;
    char val;
    UT_hash_handle hh;
};

bool isIsomorphic(char* s, char* t) {
    struct HashTable* hashtable = NULL;
    int len = strlen(s);
    int tlen = strlen(t);
    if(len!=tlen) {
        return false;
    }
    for (int i=0;i<len;i++) {
        char x = s[i], y = t[i];
        struct HashTable *tmp1, *tmp2;
        HASH_FIND(hh, hashtable, &x, sizeof(char), tmp1);
        if (tmp1 != NULL) {
            if (tmp1->val != y) {
                return false;
            }
        } else {
            tmp1 = malloc(sizeof(struct HashTable));
            tmp1->key = x;
            tmp1->val = y;
            for(int j=0;j<i;j++) {
                char z = s[j];
                HASH_FIND(hh, hashtable, &z, sizeof(char), tmp2);
                if(tmp2->val==y) {
                    return false;
                }
            }
            HASH_ADD(hh, hashtable, key, sizeof(char), tmp1);
        }
    }
    return true;
}

// @lc code=end

=======
/*
 * @lc app=leetcode id=205 lang=c
 *
 * [205] Isomorphic Strings
 */

// @lc code=start

// By hashtable
struct HashTable {
    char key;
    char val;
    UT_hash_handle hh;
};

bool isIsomorphic(char* s, char* t) {
    struct HashTable* hashtable = NULL;
    int len = strlen(s);
    int tlen = strlen(t);
    if(len!=tlen) {
        return false;
    }
    for (int i=0;i<len;i++) {
        char x = s[i], y = t[i];
        struct HashTable *tmp1, *tmp2;
        HASH_FIND(hh, hashtable, &x, sizeof(char), tmp1);
        if (tmp1 != NULL) {
            if (tmp1->val != y) {
                return false;
            }
        } else {
            tmp1 = malloc(sizeof(struct HashTable));
            tmp1->key = x;
            tmp1->val = y;
            for(int j=0;j<i;j++) {
                char z = s[j];
                HASH_FIND(hh, hashtable, &z, sizeof(char), tmp2);
                if(tmp2->val==y) {
                    return false;
                }
            }
            HASH_ADD(hh, hashtable, key, sizeof(char), tmp1);
        }
    }
    return true;
}

// @lc code=end

>>>>>>> 6861f1229a47360993e49170b9b1be7c1dd4f215
